### ISSUE TITLE
ERROR: bash: bash_execute_unix_command: cannot find keymap for command [ANSWER: due to Bash 5.1.* bug]

### ISSUE THREAD

**Issue by HenrikBengtsson** [HenrikBengtsson/shellcheck-repl#21]
# Issue

```sh
$ export SHELLCHECK_REPL_ACTION=enable; . shellcheck-repl.bash 
```

Then

```sh
$ export SHELLCHECK_REPL_ACTION=enable; . shellcheck-repl.bash 
echo $abc
     ^--^ SC2154 (warning): abc is referenced but not assigned.
     ^--^ SC2086 (info): Double quote to prevent globbing and word splitting.

To skip a check, add its SC number to 'SHELLCHECK_REPL_EXCLUDE', e.g.

  export SHELLCHECK_REPL_EXCLUDE="${SHELLCHECK_REPL_EXCLUDE},4038"

Currently, SHELLCHECK_REPL_EXCLUDE=1001,1090,1091,1113,2034,2155,2164

To skip ShellCheck validation for this call, append two spaces

$ echo $abc
bash: bash_execute_unix_command: cannot find keymap for command
$ echo $abc
```

# Session info

```sh
$ export SHELLCHECK_REPL_ACTION=sessioninfo; . shellcheck-repl.bash 
ShellCheck REPL: 0.4.0-9004
ShellCheck: 0.9.0
Bash: 5.1.16(1)-release
```

# References

* https://lists.gnu.org/archive/html/bug-bash/2021-04/msg00137.html 
  * https://lists.gnu.org/archive/html/bug-bash/2021-04/msg00136.html
* https://github.com/bminor/bash

**Comment by HenrikBengtsson** [comment]
Next step: See if this happens only for specific versions of Bash.  If so, then the likely solution will be some type of workaround for those version, especially if it is a bug that has been fixed in Bash.

**Comment by HenrikBengtsson** [comment]
Tests on Ubuntu 20.04 show it works there with:

```sh
$ export SHELLCHECK_REPL_ACTION=sessioninfo; . shellcheck-repl.bash 
ShellCheck REPL: 0.4.0-9004
ShellCheck: 0.7.0
Bash: 5.0.17(1)-release
```

So, it could be a bug that was introduces on Bash (> 5.0.17 & <= 5.1.16).

The thread ['Mysterious error after rebinding keys in Bash 5.1.4'](https://lists.gnu.org/archive/html/bug-bash/2021-04/msg00135.html) from 2021-04-16, reports the same error message for Bash 5.1.4, and they say it wasn't there in Bash 5.0.17.

From inspecting the Bash maintaier's patch of `builtins/bind.def` in <https://lists.gnu.org/archive/html/bug-bash/2021-04/msg00136.html>:

```diff
From d3cc2edb8cd37cc893dfafd3a32e10306b415929 Mon Sep 17 00:00:00 2001
From: Koichi Murase <myoga.murase@gmail.com>
Date: Fri, 16 Apr 2021 23:03:26 +0900
Subject: [PATCH] fix incorrect unbind_unix_command on overwriting keybindings

---
 builtins/bind.def | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/builtins/bind.def b/builtins/bind.def
index ccfc08dc..8ba226d3 100644
--- a/builtins/bind.def
+++ b/builtins/bind.def
@@ -139,7 +139,7 @@ bind_builtin (list)
 
   rl_outstream = stdout;
 
-  reset_internal_getopt ();  
+  reset_internal_getopt ();
   while ((opt = internal_getopt (list, "lvpVPsSXf:q:u:m:r:x:")) != -1)
     {
       switch (opt)
@@ -290,7 +290,7 @@ bind_builtin (list)
 
       if (nlen < olen)	/* fewer bind -x bindings */
 	for (d = olen - nlen, i = 0; i < olen && d > 0; i++)
-	  if (nlen == 0 || strvec_search (nbindings, obindings[i]) >= 0)
+	  if (nlen == 0 || strvec_search (nbindings, obindings[i]) < 0)
 	    {
 	      unbind_unix_command (obindings[i]);
 	      d--;
-- 
2.21.3
```

and looking at the git blame (https://github.com/bminor/bash/blame/ec8113b9861375e4e17b3307372569d429dec814/builtins/bind.def#L293), it looks like this is fixed in Bash (>= 5.2).  Look further back in git blame (https://github.com/bminor/bash/blame/9439ce094c9aa7557a9d53ac7b412a23aa66e36b/builtins/bind.def), it looks like this bug was introduced in Bash 5.1, which is also when `unbind_unix_command()` was introduced.

**Comment by HenrikBengtsson** [comment]
Testing by installing Bash from source on Ubuntu 22.04, gives that the bug is there in only Bash 5.1.*, but not before or after in Bash 5.2.  Summary:

* OK Bash: 5.0.17(1)-release
* ERROR Bash: 5.1.0(1)-release
* ERROR Bash: 5.1.16(1)-release
* OK Bash: 5.2.0(1)-release

Conclusion: Most likely a bug in Bash (>= 5.1.0 & < 5.2.0). Need to figure out a workaround for those versions.

**Comment by HenrikBengtsson** [comment]
For a starter, ShellCheck REPL will now detect the buggy Bash version and produce a warning:

```
WARNING: ShellCheck REPL does not work properly, because of a bug in Bash (>= 5.1 & < 5.2). The plan
is to find a workaround for this bug. See for https://github.com/HenrikBengtsson/shellcheck-repl/issues/21 
details and progress [shellcheck-repl 0.4.0-9005; bash 5.1.16(1)-release]
```

**Comment by HenrikBengtsson** [comment]
Example of Bash 5.1.x bug from <https://lists.gnu.org/archive/html/bug-bash/2021-04/msg00135.html>:

```sh
bash-5.1$ bind -m emacs -x '"\C-r": echo one'
bash-5.1$ bind -X
"\C-r": "echo one"
```

```sh
bash-5.1$ bind -m emacs -x '"\C-t": echo two'
bash-5.1$ bind -X
"\C-r": "echo one"
"\C-t": "echo two"
```

So far so good, but if we redefine the latter, the bug kicks in and we lose the first biniding:

```sh
bash-5.1$ bind -m emacs '"\C-t": transpose-chars'
bash-5.1$ bind -X
"\C-t": "echo two"
```


It looks like we can avoid the bug if we remove the binding first, instead of overriding it, e.g.

```sh
bash-5.1$ bind -m emacs -x '"\C-r": echo one'
bash-5.1$ bind -m emacs -x '"\C-t": echo two'
bash-5.1$ bind -X
"\C-r": "echo one"
"\C-t": "echo two"
```

```sh
bash-5.1$ bind -r "\C-t"
bash-5.1$ bind -m emacs -x '"\C-t": echo three'
bash-5.1$ bind -X
"\C-r": "echo one"
"\C-t": "echo three"
```


More over, after doing this _once_, it looks like it is safe to override an existing binding, e.g.

```r
bash-5.1$ bind -m emacs -x '"\C-t": echo two'
bash-5.1$ bind -X
"\C-r": "echo one"
"\C-t": "echo two"

bash-5.1$ bind -m emacs -x '"\C-t": echo three'
bash-5.1$ bind -X
"\C-r": "echo one"
"\C-t": "echo three"
```

and so on.

**Comment by HenrikBengtsson** [comment]
Fixed on shellcheck-repl 0.4.2 (2023-02-15)
